Add multigrid solver

nilsvu opened the pull request:
## Proposed changes

Adds the multigrid solver, which is the last remaining component of the elliptic solver. It works by solving the equations on successively coarser grids.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add child_size function in multiple dimensions

Also register the test with CMake, which must have slipped through
before.

nilsvu committed:
Support array sections in GMRES linear solver

nilsvu committed:
Support exchanging element allocator in elliptic DgElementArray

The multigrid solver will provide a different strategy of allocating
elements, because it needs to cover the domain multiple times at
different refinement levels.

nilsvu committed:
Add ObservationKey tag

nilsvu committed:
Support section observations in async linear solvers

nilsvu commented:
The first few commits that add support for multigrid in the other parallel linear solvers could be factored in a separate PR, but I kept them here for now so they are better tested by the multigrid tests.

nilsdeppe commented on src/Domain/Structure/ChildSize.hpp:
```diff
@@ -3,10 +3,14 @@
 #include "NumericalAlgorithms/Spectral/Projection.hpp"
 
 namespace domain {
+// @{
```
this conflicts with #3226

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Gmres/ElementActions.hpp:
```diff
@@ -36,15 +39,40 @@ namespace LinearSolver::gmres::detail {
+  if constexpr (std::is_same_v<ArraySectionIdTag, void>) {
+    return Parallel::no_section();
+  } else {
+    auto& section = db::get_mutable_reference<
```
are you getting a reference to a pointer here or why is `section` being dereferenced below. It looks like this is a `auto * const`? If not, a comment would be useful. Even better, write out the concrete type. I think the return type can also be a concrete type

nilsvu replied:
it's a reference to an `optional`. I made the type explicit. I left the return type as `auto&` for now because it's fairly easy to see what it is now.

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp:
```diff
@@ -77,13 +77,25 @@ namespace LinearSolver::gmres {
+ * \par Array sections
+ * This linear solver supports running over a subset of the elements in the
+ * array parallel component (see `Parallel::Section`). Set the
+ * `ArraySectionIdTag` template parameter to restrict the solver to elements in
```
I'm confused by the statement that they are only in one section based on the template parameter. How does that work with multigrid? I assumed elements were in one section based on their refinement level (and grid index), which is a runtime decision, right?

nilsvu replied:
Multigrid creates two sets of sections: One associated with `multigrid::Tags::MultigridLevel` and one with `multigrid::Tags::IsFinestLevel` (note that these tags are not actually necessarily in the DataBox, only the `Parallel::Tags::Section<ParallelComponent, SectionIdTag>` is in the DataBox). The GMRES algorithm runs only on the finest level, and the preconditioners run on all levels.

nilsdeppe replied:
Thanks for explaining that :)

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/AsynchronousSolvers/ElementActions.hpp:
```diff
@@ -163,24 +190,61 @@ struct InitializeElement {
+    // to a section with this tag. In the latter case, no observations will
+    // ever be contributed.
+    const std::string section_observation_key =
+        get_section_observation_key<ArraySectionIdTag>(box).value_or("__none");
```
What about replacing `__none` with `DoNotPerformReduction`? While not a C++ entity, `__` prefixes in C++ are reserved for compilers and STL, so I'm a little worried that such a string will encourage a variable named `__none` :Shrug:

nilsvu replied:
I went with "Unused" now. My intention was to make sure this special value doesn't conflict with anything that developers might use as a section key, but "Unused" is probably pretty safe. Added an assert to be sure.

nilsdeppe commented on src/Elliptic/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -3,24 +3,66 @@
 
 namespace elliptic {
+
+/// A `Parallel::protocols::ArrayElementsAllocator` that creates array elements
```
I feel like this documentation is missing the key information on _how_ the domain is covered, no?

nilsdeppe reviewed (dismissed):
I didn't review the last commit because reviewing a 2000 line commit is a bit much... Even if the code is in a single PR, code it at least be split into smaller, easier to review commits? I see things like FileTestDefs.sh is changed and there are tags files, etc.

nilsvu committed:
Add tags for multigrid solver

nilsvu committed:
Add elements allocator for multigrid

nilsvu committed:
Clean up volume data files from linear solver tests

Plus some minor fixes.

nilsvu commented:
@nilsdeppe thanks for your review! I pushed a few fixups, and broke the last commit up in four. Hopefully this makes it a bit better :)

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp:
```diff
@@ -0,0 +1,205 @@
+
+namespace LinearSolver::multigrid {
+
+namespace OptionTags {
```
Should this be documented? I'm not sure we have a policy for option tags namespaces or the tags themselves

nilsvu replied:
Barely any option tags namespace or option tags themselves are currently documented. We should figure out a good way to do that, but this PR is probably not the place for that :smile: (The thing with option tags is that the docs should really be in their help string)

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp:
```diff
@@ -0,0 +1,205 @@
+
+}  // namespace OptionTags
+
+namespace Tags {
```
This should be documented

nilsvu replied:
indeed :)

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp:
```diff
@@ -0,0 +1,205 @@
+    return "ObservationId(" + Options::name<OptionsGroup>() + ")";
+  }
+};
+// @{
```
group syntax

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Multigrid/ElementsAllocator.hpp:
```diff
@@ -0,0 +1,181 @@
+ * - `Parallel::Tags::Section<ElementArray,
+ *   LinearSolver::multigrid::Tags::MultigridLevel>`: One section per grid. All
+ *   elements are part of a section with this tag.
+ * - Parallel::Tags::Section<ElementArray,
```
missing backtick before `Parallel`, currently the code highlighting is wrong

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp:
```diff
@@ -0,0 +1,205 @@
+  using type = Options::Auto<size_t>;
+  static constexpr Options::String help =
+      "Maximum number of levels in the multigrid hierarchy. Includes the "
+      "finest grid, i.e. set to '1' to disable multigrids. Set to 'Auto' to "
```
Does `lower_bound` work with `Options::Auto` to enforce the bound you have here?

nilsvu replied:
It doesn't :( But there's a validation check in `ElementsAllocator` that I had added for that reason.

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Multigrid/ElementsAllocator.hpp:
```diff
@@ -0,0 +1,181 @@
+        element_array(element_id)
+            .insert(global_cache, initialization_items, target_proc);
+      }
+      Parallel::printf(
```
should this print be behind some verbosity flag?

nilsvu replied:
I had decided against it because it prints only a few lines at the very start of the simulation, and it's very important information. It should be above `Verbosity::Quiet` (which prints the residual norm after every iteration).

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Actions/RestrictFields.hpp:
```diff
@@ -0,0 +1,221 @@
+    if constexpr (not std::is_same_v<FieldsAreMassiveTag, void>) {
+      massive = db::get<FieldsAreMassiveTag>(box);
+    }
+    tuples::TaggedTuple<ReceiveTags...> restricted_fields{};
```
well the `ReceiveTags` be non-Tensor<DataVector> in the future? If not, you might want to consider using a `Variables` instead. I think you'd save not just on memory allocations but would also get better throughput on the matrix operations (`apply_matrices` does 2 memory allocations per call)

nilsvu replied:
The `ReceiveTags` will actually hold `Variables` :) The action supports restricting _multiple_ Variables. Combining the Variables into one could possibly improve performance, but this currently only happens ~5 times in a simulation (per element per multigrid-level), namely once per non-linear solver iteration.

I added a note to the docs on the requirements for the `ReceiveTags` (they must support `apply_matrices` and addition). Thanks for pointing this out!

nilsdeppe replied:
Okay, sounds good. Thanks!

nilsdeppe reviewed (commented):
I only made it up to `ElementActions.hpp`. The last commit is still pretty big so it'll take me a few goes to get through it...

nilsdeppe reviewed (commented):
LGTM, please go ahead and squash

moxcodes commented on tests/Unit/ParallelAlgorithms/LinearSolver/Gmres/Test_ElementActions.cpp:
```diff
@@ -68,13 +68,13 @@ struct ElementArray {
               LinearSolver::gmres::detail::NormalizeInitialOperand<
                   fields_tag, DummyOptionsGroup, Preconditioned,
-                  DummyOptionsGroup>,
+                  DummyOptionsGroup, void>,
```
Are there any tests for the case where there is a section being used for these actions? There's some nontrivial jumping between actions in the branches that skip elements that aren't in the present section, and it might be a good idea to have tests to verify that the logic is correct.

nilsvu replied:
It would be nice to add support for array sections to the action testing framework. That's a bit more involved though, because there's some charm-related functionality missing (e.g. can't currently instantiate `Parallel::charm_types_with_parameters` with a mock component). I think the steps for extending the action testing framework are: (1) add support for reductions, (2) add support for more low-level charm things like a `cbase`  alias etc, (3) add support for sections and section reductions. I opened https://github.com/sxs-collaboration/spectre/issues/3368 to keep track. I've been meaning to do (1), but it didn't make it up my priority list so far. The main reason is that I haven't found the linear-solver action-tests (that would benefit from reductions support) particularly useful when debugging because the actions are just chunks of the full algorithm, and the particular chunking is an implementation detail that I've had to adjust sometimes. The full-algorithm executable tests have been a lot more useful for the parallel linear solvers specifically.

To answer your question, the sectioned GMRES algorithm is tested by `Test_MultigridPreconditionedGmresAlgorithm` (added in the last commit).

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp:
```diff
@@ -0,0 +1,206 @@
+#include <array>
+#include <cstddef>
+#include <optional>
+#include <unordered_set>
```
missing `<string>`,  not sure what `Matrix` is used for in this file.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Multigrid/ElementsAllocator.hpp:
```diff
@@ -0,0 +1,181 @@
+
+  template <typename ElementArray, typename Metavariables,
+            typename... InitializationTags>
+  static void apply(Parallel::CProxy_GlobalCache<Metavariables>& global_cache,
```
I'd ask for a unit test for this, because verifying the logic for the multigrid allocation seems valuable, but I think we can't test anything that takes a `CProxy` using the component mocking in the unit tests framework.
If you can see a direct way of constructing a test, then I think that would be an improvement, but the only thing I can see at the moment would be to assemble an executable test that constructs the array, then verifies that it has the right structure -- maybe by performing a reduction? That seems pretty involved, and if we can't think of anything simpler, then I'd suggest just making an issue that currently none of the array allocation routines have unit tests and come back to it another time.

nilsvu replied:
yea it would be nice to unit-test the array allocations. I opened https://github.com/sxs-collaboration/spectre/issues/3373 to keep track.

This particular code is tested in the `Test_MultigridAlgorithm` executable test.

moxcodes commented on tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp:
```diff
@@ -343,6 +343,11 @@ struct CleanOutput {
       ERROR("Expected reductions file '" << reductions_file_name
                                          << "' does not exist");
     }
+    const auto& volume_file_name =
```
Do these volume files cause the tests to fail if they are not cleaned up? If so, it's probably best to check them and clean them up at the start of the test so that if a test is cancelled or crashes before it gets to this part, it can still succeed on subsequent runs.

nilsvu replied:
The `CleanOutput` action runs at the start _and_ at the end of the test, so the volume files are cleaned up already :)

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Multigrid/ElementActions.hpp:
```diff
@@ -0,0 +1,415 @@
+// post-smoother is skipped on the coarsest level, it directly sends the
+// solution of the pre-smoother to the finer grid, thus kicking off the
+// "ascending" branch of the V-cycle.
+template <typename FieldsTag, typename OptionsGroup, typename SourceTag>
```
Why are these actions organized differently from the send/receive going from the finer grid to the coarser one? The other set was constructed more generically, got their own file, and have publicly visible documentation.

nilsvu replied:
The restriction actions are exposed publicly because an executable may need to restrict data from the finest grid down the multigrid hierarchy, e.g. variable background fields that get updated in every non-linear solver step. There's no need for executables to interpolate anything back up the multigrid hierarchy, so the prolongation operations are private to the multigrid algorithm and more tightly integrated into the algorithm.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Multigrid.hpp:
```diff
@@ -3,6 +3,147 @@
+ *
+ * This linear solver iteratively corrects an initial guess \f$x_0\f$ by
+ * restricting the residual \f$b - Ax\f$ to a series of coarser grids, solving
+ * for a correction on the coarser grids, and then prolongating the correction
```
perhaps include the parenthetical `prolongating (interpolating)`, as I think the term 'prolongation' is not often used outside of describing elliptic solves and connecting the process with interpolation may make the documentation easier for newcomers to read.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Multigrid/Multigrid.hpp:
```diff
@@ -3,6 +3,147 @@
+ * operators.
+ *
+ * \par Smoother
+ * On every level of the grid hierarchy the multigrid solver relies on a
```
From reading the documentation and the struct, it's not clear to me how one uses a choice of smoother with this linear solver -- there's no template for it, and apart from the pre- and post- smoother actions, there doesn't appear to be any place it fits into the action list.

The mathematical documentation here is very good, but it might be helpful to include some more details on how the class is intended to be used with the other code objects when assembling an executable file or similar. Or if that documentation already exists elsewhere, it would be nice to reference it somewhere here.

nilsvu replied:
Added some snippets, hopefully that makes it clearer.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Multigrid/ObserveVolumeData.hpp:
```diff
@@ -0,0 +1,126 @@
+#include <cstddef>
+#include <string>
+#include <tuple>
+#include <utility>
```
missing `<type_traits>`, `Utilities/Gsl.hpp`

moxcodes commented on tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/Multigrid/Helpers.hpp:
```diff
@@ -0,0 +1,252 @@
+
+    typename OperandTag::type operator_applied_to_operand{
+        number_of_grid_points * number_of_elements};
+    dgemv_('N', linear_operator.rows(), linear_operator.columns(), 1,
```
Can you use `ApplyMatrices` for this?

nilsvu replied:
I had to add a commit that makes `Matrix` a type alias (instead of a derived class) for this to work with both option-construction and `apply_matrices`. I pushed the commit here and also to #3374.

nilsvu replied:
Actually, this leads down a whole rabbit hole related to Blaze-interoperability. If you don't mind I'll leave the explicit blas call here for now and deal with the Blaze issues separately.

moxcodes replied:
Okay, you can leave it as-is, then, just add a code comment explaining the reason you can't use `apply_matrices`.

moxcodes commented on tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/Multigrid/Helpers.hpp:
```diff
@@ -0,0 +1,252 @@
+        const Parallel::GlobalCache<Metavariables>& /*cache*/,
+        const ElementId<1>& element_id, const ActionList /*meta*/,
+        const ParallelComponent* const /*meta*/) noexcept {
+    const size_t mg_lev = element_id.grid_index();
```
maybe `multigrid` instead of `mg` for more readability.

moxcodes commented on tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/Multigrid/Helpers.hpp:
```diff
@@ -0,0 +1,252 @@
+  static void apply(
+      db::DataBox<DbTagsList>& box, Parallel::GlobalCache<Metavariables>& cache,
+      const ElementId<1>& element_id,
+      const Variables<tmpl::list<ScalarFieldOperandTag>>& Ap_global_data,
```
What does the `Ap` prefix mean in this name?

nilsvu replied:
It means the linear operator `A` applied to the operand `p`, i.e. `A \cdot p`. Changed to `operator_applied_to_operand`.

moxcodes commented on tests/Unit/ParallelAlgorithms/LinearSolver/Multigrid/Test_MultigridAlgorithm.cpp:
```diff
@@ -0,0 +1,128 @@
+// Distributed under the MIT License.
```
It could be nice to have unit tests using the action-testing framework for any of the actions that can be tested on their own, so that any problems or changes that arise in particular actions can be more easily found and fixed.

nilsvu replied:
As mentioned above https://github.com/sxs-collaboration/spectre/pull/3249#discussion_r672615525 I didn't find the action tests for the internal actions of the parallel linear solvers particularly useful so far. I did add a fixup with action-tests for the public restriction actions though, so thanks for pointing this out!

nilsvu commented:
@moxcodes fixups posted. Cheers!

nilsvu commented:
@moxcodes Could you take a look at the fixups before I rebase this?

moxcodes reviewed (commented):
I'm happy with the fixups, go ahead and rebase and squash.

nilsvu committed:
Add multigrid restrictions actions

These can be used to restrict data down the multigrid hierarchy,
which is often necessary for background fields that get updated
throughout the solve. The counterpart to the restriction, the
prolongation operation that interpolates data back up the
multigrid hierarchy, does _not_ need to be exposed publicly.

nilsvu committed:
Add parallel multigrid solver algorithm

nilsvu commented:
This is now rebased and squashed. Thanks for your reviews @nilsdeppe and @moxcodes!

@kidder could you sign off on this since @nilsdeppe is absent? He already approved it before @moxcodes's review.
